trunk/MOCLE/man/man3/Partition.3
.TH "Partition" 3 "27 Jul 2010" "Version version2.0" "MOCLE" \" -*- nroff -*-
.ad l
.nh
.SH NAME
Partition \- 
.PP
Responsible to manage the partition file into the memory.  

.SH SYNOPSIS
.br
.PP
.PP
\fC#include <Partition.h>\fP
.SS "Classes"

.in +1c
.ti -1c
.RI "class \fBCluster\fP"
.br
.RI "\fIEach element in the dataset is allocated in one cluster. \fP"
.in -1c
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBPartition\fP (string sAPathPartition, string sANamePartition)"
.br
.ti -1c
.RI "void \fBcalculateCentroid\fP (\fBDataSet\fP *pAObjDataSet)"
.br
.ti -1c
.RI "void \fBloadPartition\fP ()"
.br
.ti -1c
.RI "void \fBgenerateRandomPartition\fP (\fBDataSet\fP *pAObjDataSet)"
.br
.ti -1c
.RI "void \fBgeneratePartitionGroupCluster\fP (\fBDataSet\fP *pAObjDataSet)"
.br
.ti -1c
.RI "void \fBprintClusterLabels\fP ()"
.br
.ti -1c
.RI "void \fBprintClusters\fP ()"
.br
.ti -1c
.RI "void \fBshowCentroids\fP ()"
.br
.ti -1c
.RI "void \fBwritePartition\fP ()"
.br
.ti -1c
.RI "Cluster & \fBgetCluster\fP (int iAPos)"
.br
.ti -1c
.RI "int \fBgetNumClusters\fP ()"
.br
.ti -1c
.RI "int \fBgetNumObjects\fP ()"
.br
.ti -1c
.RI "map< string, int > \fBgetMObjects\fP ()"
.br
.ti -1c
.RI "vector< Partition::Cluster > \fBgetVObjects\fP ()"
.br
.ti -1c
.RI "int \fBgetLabelClusterOf\fP (string sAObject)"
.br
.ti -1c
.RI "vector< Partition::Cluster > \fBgetVectorObjCluster\fP ()"
.br
.ti -1c
.RI "vector< string > \fBgetObjectsInCluster\fP (int ClusterLabel)"
.br
.ti -1c
.RI "vector< double > \fBgetCentroidInCluster\fP (int ClusterLabel)"
.br
.in -1c
.SH "Detailed Description"
.PP 
Responsible to manage the partition file into the memory. 

\fBAuthor:\fP
.RS 4
Valter Henrique, Vinicius Lopes, Gustavo Henrique, Katti Facelli 
.RE
.PP
\fBSince:\fP
.RS 4
09/02/10 
.RE
.PP
\fBVersion:\fP
.RS 4
2.0 Colocar um comentário mais detalhado na classe dataset 
.RE
.PP

.PP
Definition at line 26 of file Partition.h.
.SH "Constructor & Destructor Documentation"
.PP 
.SS "Partition::Partition (string sAPathPartition, string sANamePartition)"Does nothing 
.PP
\fBParameters:\fP
.RS 4
\fIDon't\fP have any parameter 
.RE
.PP

.PP
Definition at line 10 of file Partition.cpp.
.PP
References loadPartition().
.SH "Member Function Documentation"
.PP 
.SS "void Partition::calculateCentroid (\fBDataSet\fP * pAObjDataSet)"Calculate the centroid of the cluster 
.PP
\fBParameters:\fP
.RS 4
\fIDon't\fP have 
.RE
.PP
\fBSee also:\fP
.RS 4
calculateCentroid(DataSet &obADataSet) 
.RE
.PP

.PP
Definition at line 18 of file Partition.cpp.
.SS "void Partition::generatePartitionGroupCluster (\fBDataSet\fP * pAObjDataSet)"Generates a random partition, grouping the objects in the same cluster,receiving as parameter a point to the object of the dataset 
.PP
\fBParameters:\fP
.RS 4
\fI*obADataSet\fP 
.RE
.PP
\fBReturns:\fP
.RS 4
Don't have 
.RE
.PP

.PP
Definition at line 227 of file Partition.cpp.
.SS "void Partition::generateRandomPartition (\fBDataSet\fP * pAObjDataSet)"Generates a random partition, receiving as parameter a point to the object of the dataset 
.PP
\fBParameters:\fP
.RS 4
\fI*obADataSet\fP 
.RE
.PP
\fBReturns:\fP
.RS 4
Don't have 
.RE
.PP

.PP
Definition at line 174 of file Partition.cpp.
.PP
References DataSet::getNumberOfObjects().
.SS "vector< double > Partition::getCentroidInCluster (int ClusterLabel)"Returns the centroid of Cluster 
.PP
\fBParameters:\fP
.RS 4
\fICluster\fP Label 
.RE
.PP
\fBReturns:\fP
.RS 4
vector<double> 
.RE
.PP

.PP
Definition at line 27 of file Partition.cpp.
.PP
Referenced by Deviation::calculate().
.SS "Partition::Cluster & Partition::getCluster (int iAPos)"Returns a specific cluster passing as argument the position where he is allocated 
.PP
\fBParameters:\fP
.RS 4
\fIint\fP iPos 
.RE
.PP

.PP
Definition at line 153 of file Partition.cpp.
.PP
Referenced by CRIndex::calculate().
.SS "int Partition::getLabelClusterOf (string sAObject)"Returns the cluster's label of parameter sObject 
.PP
\fBParameters:\fP
.RS 4
\fIsObject\fP indicated the object label 
.RE
.PP
\fBReturns:\fP
.RS 4
Cluster's label 
.RE
.PP

.PP
Definition at line 35 of file Partition.cpp.
.PP
Referenced by Connectivity::calculate().
.SS "map< string, int > Partition::getMObjects ()"Returns mObjects 
.PP
\fBParameters:\fP
.RS 4
\fIDon't\fP have 
.RE
.PP
\fBReturns:\fP
.RS 4
map<string, int> mObjects 
.RE
.PP

.PP
Definition at line 165 of file Partition.cpp.
.SS "int Partition::getNumClusters ()"Returns the number of cluster existing cluster on the partition, loaded in the \fBloadPartition()\fP 
.PP
\fBParameters:\fP
.RS 4
\fIDon't\fP have 
.RE
.PP
\fBReturns:\fP
.RS 4
iNumClusters 
.RE
.PP
\fBSee also:\fP
.RS 4
loadPartition(fs::path pathAPartition), iNumClusters 
.RE
.PP

.PP
Definition at line 157 of file Partition.cpp.
.PP
Referenced by CRIndex::calculate().
.SS "int Partition::getNumObjects ()"Returns the number of existing objects on the partition, loaded in the \fBloadPartition()\fP 
.PP
\fBParameters:\fP
.RS 4
\fIDon't\fP have 
.RE
.PP
\fBSee also:\fP
.RS 4
loadPartition(fs::path pathAPartition) 
.RE
.PP

.PP
Definition at line 161 of file Partition.cpp.
.PP
Referenced by CRIndex::calculate().
.SS "vector< string > Partition::getObjectsInCluster (int ClusterLabel)"Returns the rotule objects in ClusterLabel 
.PP
\fBParameters:\fP
.RS 4
\fICluster\fP Label 
.RE
.PP
\fBReturns:\fP
.RS 4
vector<string> 
.RE
.PP

.PP
Definition at line 23 of file Partition.cpp.
.PP
Referenced by Silhouette::calculate(), and Deviation::calculate().
.SS "vector< Partition::Cluster > Partition::getVectorObjCluster ()"Returns the vector all clusters of partition 
.PP
\fBParameters:\fP
.RS 4
\fIDon't\fP have 
.RE
.PP
\fBReturns:\fP
.RS 4
vector<Cluster> 
.RE
.PP

.PP
Definition at line 31 of file Partition.cpp.
.PP
Referenced by Silhouette::calculate(), and Deviation::calculate().
.SS "vector< Partition::Cluster > Partition::getVObjects ()"Returns vectorObjects, that contains objects of cluster type 
.PP
\fBParameters:\fP
.RS 4
\fIDon't\fP have 
.RE
.PP
\fBReturns:\fP
.RS 4
vector<Cluster> vObCluster 
.RE
.PP

.PP
Definition at line 169 of file Partition.cpp.
.SS "void Partition::loadPartition ()"charge to the partition in the memory, passing in the argument the path to load the partition in memory, passing as argument the path to the partition 
.PP
\fBParameters:\fP
.RS 4
\fIpathAPartition\fP 
.RE
.PP
\fBReturns:\fP
.RS 4
Don't have. 
.RE
.PP

.PP
Definition at line 53 of file Partition.cpp.
.PP
Referenced by Partition().
.SS "void Partition::printClusterLabels ()"Print the labels of the partition 
.PP
Definition at line 283 of file Partition.cpp.
.SS "void Partition::writePartition ()"write the partition in disk, passing in the argument the path where the user want to save 
.PP
\fBParameters:\fP
.RS 4
\fIpathSaveIn\fP 
.RE
.PP
\fBReturns:\fP
.RS 4
Don't have. 
.RE
.PP

.PP
Definition at line 116 of file Partition.cpp.

.SH "Author"
.PP 
Generated automatically by Doxygen for MOCLE from the source code.


trunk/MOCLE/Partition.cpp
/*
 * Partition.cpp
 *
 *  Created on: 09/02/2010
 *      Author: Valter
 */

#include "Partition.h"

Partition::Partition(string sAPathPartition, string sANamePartition) {
	sPathPartition = sAPathPartition;
	sNamePartition = sANamePartition;
	loadPartition();
}

Partition::~Partition() {}

void Partition::calculateCentroid(DataSet *pAObjDataSet) {
	for (int i = 0; i < (int)vectorObjCluster.size(); i++)
		vectorObjCluster[i].calculateCentroid(pAObjDataSet);
}

vector<string> Partition::getObjectsInCluster(int ClusterLabel){
	return vectorObjCluster[ClusterLabel].getVectorObjects();
}

vector<double> Partition::getCentroidInCluster(int ClusterLabel){
	return vectorObjCluster[ClusterLabel].getVectorCentroids();
}

vector<Partition::Cluster> Partition::getVectorObjCluster(){
	return vectorObjCluster;
}

int Partition::getLabelClusterOf(string sAObject){
	vector<string>::iterator itVectorObjects;
	vector<string> vectorObjectsOnCluster;
	itCluster it;

	for(it = begin(); it != end(); ++it){
		vectorObjectsOnCluster = it->getVectorObjects();

		for(itVectorObjects = vectorObjectsOnCluster.begin(); itVectorObjects != vectorObjectsOnCluster.end(); itVectorObjects++){

			if( (*itVectorObjects) == sAObject){
				return it->getLabel();
			}
		}
	}
	return -1; //VERIFICAR SE A CONSTANTE DE ERRO SERA -1
}

void Partition::loadPartition() {
	Partition::Cluster *pObCluster;
	string sTemp;
	string sObject;
	string sPath = sPathPartition + sNamePartition;
	int iCluster;
	int i;

	ifstream in(sPath.c_str());

	// define what gonna be the tokenizer character
	char_separator<char> sep("\t");

	while (getline(in, sTemp)) {

		// tokenize the string sTemp that's the line read in the file informed in 'pathAPartition'
		tokenizer<char_separator<char> > tokens(sTemp, sep);

		for (tokenizer<char_separator<char> >::iterator itTokens = tokens.begin(); itTokens != tokens.end(); ++itTokens) {

			// putting the ID of the element in the memory
			// mapVector[sIndexLine].push_back(dNumber);
			sObject = *itTokens;
			++itTokens;

			(istringstream(*itTokens) >> iCluster);
			mapObjects[sObject] = iCluster;

			i = 0;
			while (i < (int)vectorObjCluster.size() && (int)vectorObjCluster[i].getLabel()!= iCluster) {
				i++;
			}

			if (i == (int)vectorObjCluster.size()) {
				pObCluster = new Cluster(iCluster, sObject);
				vectorObjCluster.push_back(*pObCluster);
				delete pObCluster;
			} else
				vectorObjCluster[i].addObject(sObject);
		}
	}

	iNumObjects = mapObjects.size();
	iNumClusters =  vectorObjCluster.size();

}

void Partition::showCentroids(){

	vector<double> vectorCentroid;
	vector<double>::iterator itVectorCentroid;

	for (int i = 0; i < iNumClusters; i++){

		 vectorCentroid = vectorObjCluster[i].getVectorCentroids();

		for (itVectorCentroid = vectorCentroid.begin(); itVectorCentroid != vectorCentroid.end(); ++itVectorCentroid){
			cout << "Cluster (" << i << ") | Centroid >> " << *itVectorCentroid << endl;
		}
	}

}

void Partition::writePartition() {

	string sFolder;
	string sNamePartition;

	cout << "Inform the name of the folder that you want save your partition" << endl;
	cout << ">> ";
	cin >> sFolder;

	// # ficando o caminho para a nova pasta do tipo da partition
	cout << "New folder in : " << sPathPartition + sFolder << endl;

	//fs::path pathPartition(sFolder);

	// # criando o diretório especificado pelo usuário
	fs::create_directories(sPathPartition+sFolder);

	cout << "Inform the name of the file that gonna be your partition" << endl;
	cout << ">> ";
	cin >> sNamePartition;

	string sPath = sPathPartition + sFolder + "/" + sNamePartition + ".clu";

	cout << "sPath no partition >> " << sPath;

	ofstream out(sPath.c_str());

	map <string, int>::iterator itMapObjects;

	for(itMapObjects = mapObjects.begin(); itMapObjects != mapObjects.end(); ++itMapObjects){
		out << itMapObjects->first << "\t" << itMapObjects->second << "\n";
	}

	out.close();
}


Partition::Cluster & Partition::getCluster(int iPos) {
	return vectorObjCluster[iPos];
}

int Partition::getNumClusters() {
	return iNumClusters;
}

int Partition::getNumObjects() {
	return iNumObjects;
}

map<string, int> Partition::getMObjects(){
	return mapObjects;
}

vector<Partition::Cluster> Partition::getVObjects(){
	return vectorObjCluster;
}

// in this method in generated a random partition, don't follow any standard
void Partition::generateRandomPartition(DataSet *pAObjDataSet) {

	time_t rawtime;
	struct tm * timeinfo;
	time (&rawtime);
	timeinfo = localtime(&rawtime);

	string sMonth = boost::lexical_cast<string>(timeinfo->tm_mon+1);
	string sDay   = boost::lexical_cast<string>(timeinfo->tm_mday);
	string sYear  = boost::lexical_cast<string>(1900 + timeinfo->tm_year);

	string sHour   = boost::lexical_cast<string>(timeinfo->tm_hour);
	string sMinute = boost::lexical_cast<string>(timeinfo->tm_min);
	string sSecond = boost::lexical_cast<string>(timeinfo->tm_sec);

	string sFolder = sDay + "|" + sMonth + "|" + sYear + " " + sHour + ":" + sMinute + ":" + sSecond;
	string sName;

	int iObjects = pAObjDataSet->getNumberOfObjects();
	int iNumClusters = -1;

	fs::create_directories(sPathPartition + sFolder);

	cout << "Inform the NAME of the new partition" << endl;
	cout << ">> ";
	cin >> sName;

	/*
	while (iObjects <= 0 ){
		cout << "Inform the number of OBJECTS that you want in the new partition" << endl;
		cout << ">> ";
		cin >> iObjects;
	}
	*/

	while (iNumClusters <= 0){
		cout << "Inform the number of CLUSTERS that you want in the new partition" << endl;
		cout << ">> ";
		cin >> iNumClusters;
	}

	string sNewPartition = sPathPartition + sFolder + "/" + sName + ".clu";

	ofstream out( sNewPartition.c_str());

	for(int i = 0; i < iObjects; i++)	{
		out << "Element" << "\t" << rand()%iNumClusters + 1 << endl;
	}

	out.close();

}

void Partition::generatePartitionGroupCluster(DataSet *pAObjDataSet) {

	time_t rawtime;
	struct tm * timeinfo;
	time (&rawtime);
	timeinfo = localtime(&rawtime);

	string sMonth = boost::lexical_cast<string>(timeinfo->tm_mon+1);
	string sDay =   boost::lexical_cast<string>(timeinfo->tm_mday);
	string sYear =  boost::lexical_cast<string>(1900 + timeinfo->tm_year);

	string sHour =   boost::lexical_cast<string>(timeinfo->tm_hour);
	string sMinute = boost::lexical_cast<string>(timeinfo->tm_min);
	string sSecond = boost::lexical_cast<string>(timeinfo->tm_sec);

	string sFolder = sDay + "|" + sMonth + "|" + sYear + " " + sHour + ":" + sMinute + ":" + sSecond;
	string sName;

	int iObjects;
	int iNumClusters;
	int iCont = 0;
	int iCluster = 1;

	fs::create_directories(sPathPartition + sFolder);

	cout << "Provide the NAME of the new partition" << endl;
	cout << ">> ";
	cin >> sName;

	cout << "Provide the number of OBJECTS that you want in the new partition" << endl;
	cout << ">> ";
	cin >> iObjects;

	cout << "Provide the number of CLUSTERS that you want in the new partition" << endl;
	cout << ">> ";
	cin >> iNumClusters;


	string sNewPartition = sPathPartition + sFolder + "/" + sName + ".clu";

	ofstream out( sNewPartition.c_str());

	for(int i = 1; i <= iObjects; i++){
		if (iCont == (iObjects / iNumClusters)){
			iCluster++;
			iCont = 0;
		}
		iCont++;
		out << i << '\t' << iCluster << endl;
	}

	out.close();

}

// # ambos os métodos printClusterLabels e printClusters fazem a mesma coisa, porém nota-se a simplificação de implementação de um método para o outro
void Partition::printClusterLabels(){

	for (vector<Partition::Cluster>::iterator itVectorObjCluster = vectorObjCluster.begin(); itVectorObjCluster != vectorObjCluster.end(); ++ itVectorObjCluster){
		cout << itVectorObjCluster->getLabel() << ", ";
	}
}

// # ambos os métodos printClusterLabels e printClusters fazem a mesma coisa, porém nota-se a simplificação de implementação de um método para o outro
void Partition::printClusters(){
	itCluster it;

	for (it = begin(); it != end(); ++ it){
		cout << it->getLabel() << ", ";
	}

}

/******************
	Cluster
******************/

Partition::Cluster::Cluster(){}

Partition::Cluster::Cluster(int iACluster, string sAID) {
	iLabel = iACluster;
	vectorObjects.push_back(sAID);
}

vector<string> Partition::Cluster::getVectorObjects(){
	return vectorObjects;
}

void Partition::Cluster::addObject(string sAID) {
	vectorObjects.push_back(sAID);
}

void Partition::Cluster::calculateCentroid(DataSet *pAObjDataSet) {
	double dSum = 0;
	int iNumFeatures = pAObjDataSet->getNumberOfFeature();

	vector<double> vectorAtributes;

	// a mapVector that will receive the mapVector from DataSet
	map<string, vector<double> > mapVector;
	mapVector = pAObjDataSet->getMapVector();

	// analyzing all that are a feature and adding their respective values and dividing by the number of elements that are in the same cluster, and storing this value vector of centroids, which have the values of the centroids
	for (int i = 0; i < iNumFeatures; i++) {
		for (int j = 0; j < (int)vectorObjects.size(); j++) {
			vectorAtributes = mapVector.find(vectorObjects.at(j))->second;
			dSum += vectorAtributes.at(i);
		}
		vectorCentroid.push_back(dSum / vectorObjects.size());

		// zeroing to sum the next feature
		dSum = 0;
	}
}

int Partition::Cluster::getLabel() {
	return iLabel;
}

int Partition::Cluster::getNumberOfObjects() {
	return vectorObjects.size();
}

vector<double> Partition::Cluster::getVectorCentroids(){
	return vectorCentroid;
}

vector<string>::iterator Partition::Cluster::objectsItBegin() {
	return vectorObjects.begin();
}

vector<string>::iterator Partition::Cluster::objectsItEnd() {
	return vectorObjects.end();
}
